product_multi_image/models/product_template.py

```diff
@@ -22,3 +22,30 @@ class ProductTemplate(models.Model):
         related='image_main_small',
         store=False,
     )
+
+    @api.multi
+    def write(self, vals):
+        if set(['image', 'image_medium', 'image_small']) & set(vals.keys()):
+            # medium/small image is saved as large
+            # on load image will be medium/small resized on the fly
+            if 'image' in vals:
+                pass
+            elif 'image_medium' in vals:
+                vals['image'] = vals['image_medium']
+                del vals['image_medium']
+            elif 'image_small' in vals:
+                vals['image'] = vals['image_small']
+                del vals['image_small']
+        return super(ProductTemplate, self).write(vals)
+
+    @api.multi
+    def _set_multi_image_main_medium(self):
+        # on save product module resizes large image to medium
+        # medium image should not overwrite the large
+        pass
```

Comment:
I'm not sure about this inheritance break, as you are not calling super here in any case. Can you discriminate any way when to not call super? What is the use case for always inhibiting this method?

Reply:
There was an issue saving an image as the main of a product, it was **resized to small** before saving:

This happens because:
1. *product* module resizes an image to all sizes before saving it.
  https://github.com/OCA/OCB/blob/7a824c376ffe0ef6326c7c03a12e9105f9ce9f51/addons/product/models/product_template.py#L289
  https://github.com/OCA/OCB/blob/7a824c376ffe0ef6326c7c03a12e9105f9ce9f51/odoo/tools/image.py#L266
2. *product_multi_image* redefines image fields for all sizes as related from main image.
3. *base_multi_image* saves main image for all sizes in the same field:
  https://github.com/OCA/server-tools/blob/201129d9905f1312405bf4f3e08bf6e082312ac9/base_multi_image/models/owner.py#L19

Consequently the images for all sizes are saved in the same field and the last one, in this case the small one, wins :hatched_chick::trophy:

As an image is always saved for all sizes and, to not lose quality, I always save it as the largest, I fix the issue not saving it for the rest of sizes.

I can not think of a better solution... :confused:

Reply:
Have studied this also. I think the solution is good because it's an exception for products only, necessary because products have already this built-in resizing on write. As [base_multi_image mentions clearly](https://github.com/OCA/server-tools/blob/10.0/base_multi_image/models/owner.py#L53), the `_set_multi_image_main_*` family of functions is meant only as a compatibility layer. Product IMO could be an exception to this rule - the images are still saved correctly and without loss.